### PR TITLE
GH# 74: Snap to streets

### PR DESCRIPTION
**Pull request by rsarathy**

### CODE REVIEW

**Comment by rsarathy** [comment]
image
We now avoid sidewalks (namely, any ways tagged `highway=footway`) when snapping start and end locations.

File: reader-gtfs/src/main/java/com/graphhopper/gtfs/PtLocationSnapper.java [hunk 1]
```diff
@@ -47,11 +53,19 @@ public Result snapAll(List<GHLocation> locations, List<EdgeFilter> snapFilters)
...
+        final EnumEncodedValue<RoadClass> roadClassEnc = encodingManager.getEnumEncodedValue(RoadClass.KEY, RoadClass.class);
+        final EnumEncodedValue<RoadEnvironment> roadEnvEnc = encodingManager.getEnumEncodedValue(RoadEnvironment.KEY, RoadEnvironment.class);
         for (int i = 0; i < locations.size(); i++) {
             GHLocation location = locations.get(i);
             if (location instanceof GHPointLocation) {
                 GHPoint point = ((GHPointLocation) location).ghPoint;
-                final Snap closest = locationIndex.findClosest(point.lat, point.lon, snapFilters.get(i));
+                EdgeFilter mergedFilter = new SnapPreventionEdgeFilter(
+                        snapFilters.get(i), roadClassEnc, roadEnvEnc,
+                        Collections.singletonList("footway"));
```
Comment (graue) [inline reader-gtfs/src/main/java/com/graphhopper/gtfs/PtLocationSnapper.java:66]: Would we still want to snap to sidewalks for public transit directions with walking as the connecting mode? Probably, right? If so, this should test for the connecting mode. Not currently relevant but matters when we add a transit/no bike directions mode or want to merge changes upstream.

**Review by graue — APPROVED** [review]
Would it have been harder to just avoid [footway=sidewalk](https://wiki.openstreetmap.org/wiki/Tag:footway%3Dsidewalk)?

I suggested that in case there are some footways we might want to use. On the other hand, this is probably fine. Looking at the [footway](https://wiki.openstreetmap.org/wiki/Tag:highway%3Dfootway) wiki page it's mostly sidewalk-like things and (for example) the O'Shaughnessy Blvd sidepath is tagged highway=path instead, so it's NBD.

File: reader-gtfs/src/main/java/com/graphhopper/gtfs/PtLocationSnapper.java (hunk 1 above)
Comment (rsarathy) [reply reader-gtfs/src/main/java/com/graphhopper/gtfs/PtLocationSnapper.java:66]: Makes sense. Let me try and implement this at the config-level, using a `snap_preventions` field underneath our `bike2` profile.

Comment (rsarathy) [reply reader-gtfs/src/main/java/com/graphhopper/gtfs/PtLocationSnapper.java]: Done, this is now handled via our configs. We can specify `snap_preventions` as a comma-separated list.

**Comment by rsarathy** [comment]
> Would it have been harder to just avoid [footway=sidewalk](https://wiki.openstreetmap.org/wiki/Tag:footway%3Dsidewalk)?
> 
> I suggested that in case there are some footways we might want to use. On the other hand, this is probably fine. Looking at the [footway](https://wiki.openstreetmap.org/wiki/Tag:highway%3Dfootway) wiki page it's mostly sidewalk-like things and (for example) the O'Shaughnessy Blvd sidepath is tagged highway=path instead, so it's NBD.

I'm not sure right now how difficult this will be to implement, but I'll investigate it in a separate PR. I see that you opened #123, which captures this ask.

**Comment by graue** [comment]
#123 is different. That issue is about discouraging routing on sidewalks, not about preventing snapping to them for the start point of a route.

**Pull request merged**